heron/stmgr/src/cpp/manager/stateful-restorer.cpp

```diff
@@ -0,0 +1,260 @@
+  instance_connections_pending_ = true;
+  std::vector<proto::system::Instance*> instances;
+  server_->GetInstanceInfo(instances);
+  local_taskids_.clear();
+  for (auto instance : instances) {
+    local_taskids_.insert(instance->info().task_id());
+  }
+  restore_pending_ = local_taskids_;
+  get_ckpt_pending_ = local_taskids_;
+  checkpoint_id_ = _checkpoint_id;
+  restore_txid_ = _restore_txid;
+
+  // Send messages to ckpt
+  GetCheckpoints();
+
+  clientmgr_->StartConnections(_pplan);
+  if (clientmgr_->AllStMgrClientsRegistered()) {
+    // Its possible that this is really a restore while we were already in progress
+    // and there was no change in pplan. In which case there would be no new
+    // connections to restore
+    clients_connections_pending_ = false;
+  }
+  if (server_->HaveAllInstancesConnectedToUs()) {
+    instance_connections_pending_ = false;
+  }
+}
+
+void StatefulRestorer::GetCheckpoints() {
+  for (auto task_id : get_ckpt_pending_) {
+    ckptmgr_->GetInstanceState(*(server_->GetInstanceInfo(task_id)), checkpoint_id_);
+    multi_count_metric_->scope(METRIC_CKPT_REQUESTS)->incr();
+  }
+}
+
+void StatefulRestorer::HandleCheckpointState(proto::system::StatusCode _status, sp_int32 _task_id,
+                                       sp_string _checkpoint_id,
+                                       const proto::ckptmgr::InstanceStateCheckpoint& _state) {
+  LOG(INFO) << "Got InstanceState from checkpoint mgr for task " << _task_id
+            << " and checkpoint " << _state.checkpoint_id();
+  multi_count_metric_->scope(METRIC_CKPT_RESPONSES)->incr();
+  if (!in_progress_) {
+    LOG(INFO) << "Ignoring it because we are not in restore";
+    multi_count_metric_->scope(METRIC_CKPT_RESPONSES_IGNORED)->incr();
+    return;
+  }
+  if (_checkpoint_id != checkpoint_id_) {
+    LOG(INFO) << "InstanceState from checkpont mgr for a checkpoint_id "
+              << _checkpoint_id << " that is different from ours "
+              << checkpoint_id_;
+    multi_count_metric_->scope(METRIC_CKPT_RESPONSES_IGNORED)->incr();
+    return;
+  }
+  if (_status == proto::system::OK) {
+    if (_state.checkpoint_id() != checkpoint_id_) {
+      LOG(WARNING) << "Discarding state retrieved from checkpoint mgr because the checkpoint"
+                   << " id in the response does not match ours " << checkpoint_id_;
+      multi_count_metric_->scope(METRIC_CKPT_RESPONSES_IGNORED)->incr();
+      return;
+    }
+    if (server_->SendRestoreInstanceStateRequest(_task_id, _state)) {
+      multi_count_metric_->scope(METRIC_INSTANCE_RESTORE_REQUESTS)->incr();
+      get_ckpt_pending_.erase(_task_id);
+    }
+    // Note that if we are unable to send the restore instance state request(i.e. if the
+    // last call returns false, its because of connection issues and we will get back to
+    // restore anyways.
+  } else {
+    LOG(INFO) << "InstanceState from checkpont mgr contained non ok status " << _status;
+    in_progress_ = false;
```

Comment:
Why does in_progress_ get set to false if it's a non-OK status, but not if it's an invalid checkpoint id that's being ignored?